### PR TITLE
Alcor_Performance_Report

### PR DESCRIPTION
**Pull request by yanmo96**
Alcor Performance Report.
Still need to add a bit more info.
Please let me know if anything need to be changed.
Thanks

### CODE REVIEW

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 1]
```diff
@@ -0,0 +1,718 @@
+# Introduction
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Please put TOC (Table of Content) and their link to the section in the beginning of the report.

**Review by cj-chung — CHANGES_REQUESTED** [review]

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 2]
```diff
@@ -0,0 +1,718 @@
...
+To **SOLVE** this slow cleanup issue, we disabled the network cleanup in Rally's Python code, then reset our Alcor in the K8s cluster and delete Alcor's DB. From our solution, we can shorten a 50 mins long test in to just 5 mins.  
+- Second issue we are running into that cause the performance tests to last really long is during end-to-end tests with VM creation. Not only this also include the above issue for VM clean up. It also have issues during creating VM. When during Rally test, VM don't always created successfully. Rally will 300 seconds per VM if something goes wrong. And if some VM was not created successfully, Rally may very well also have issues deleting that particular VM, which will waite for another 600 seconds before time out.  
+Our solution to help with this situation are: Rebooting ACA, Nova-compute and cleanup OVS on all compute node, rebooting Nova service in the controllers. Those steps won't generate 100% success rate, but will definitely help with the situation.  
+
+# Performance Setup  
+
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Performance Test Environment Setup

**Review by cj-chung — CHANGES_REQUESTED** [review]

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 3]
```diff
@@ -0,0 +1,718 @@
...
+| Controller 3: 10.213.43.171 | Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz  | 2             | 32         | 128         |
+| Controller 4: 10.213.43.91  | Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz | 1             | 24         | 32          |
+| Controller 5: 10.213.43.94  | Intel(R) Xeon(R) CPU E5-2620 v3 @ 2.40GHz | 2             | 24         | 128         |
+
+**Our compute nodes configurations:**  
+- In total we have 2,808 vCPUs, 16,263 GB of RAM, and 64,254 GB of Storage
+
+| VCPUs (total) | RAM (total) | Local Storage (total) |
+| ------------- | ----------- | --------------------- |
+| 80            | 755.5GB     | 730GB                 |
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Please give it one more column before "VCPUs" to indicate the Number or name for computer node, such as "compute1", "compute2", etc.

**Review by cj-chung — CHANGES_REQUESTED** [review]

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+<br />
+
+### How Rally interact with OpenStack and test
+
+![alt text](images/Rally-Actions.png)
+
+Rally work along side of OpenStack, it will try to simulate the user input from OpenStack dashboard or CLI, but it will also do some direct calls to each openstack services, but with very similar workflow as organic user inputs.  
+
+# Performance Test Insights  
+
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 It's better rename this section to "Performance Test Cases and Scenarios" and briefly describe what test cases and scenarios we tested.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+| node\_manager            | 5     | 5      | 5      | 5      | 5      |
+| quota\_manager           | 5     | 5      | 5      | 5      | 5      |
+| sg\_manager              | 5     | 5      | 5      | 5      | 5      |
+| vpc\_manager             | 5     | 5      | 5      | 5      | 5      |
+| port\_manager            | 5     | 10     | 15     | 20     | 25     |
+| private\_ip\_manager     | 5     | 10     | 15     | 20     | 25     |
+| route\_manager           | 5     | 10     | 15     | 20     | 25     |
+| subnet\_manager          | 5     | 10     | 15     | 20     | 25     |
+
+## API Test, Port:
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 add a upper level section header "Performance Test Result" before "API Test, Port". Rename "API Test, Port" to "Alcor API Test for Port creation and list"

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+| run 4 | 1911.28605 | 2494.337   | 3078.50034 | 2987.68877 | 2765.557526 |
+| run 5 | 2519.02377 | 3374.84124 | 2971.57462 | 2623.8086  | 2519.896116 |
+| Avg   | 2481.3188  | 2819.13026 | 2957.79381 | 2750.94222 | 2592.338331 |
+| Max   | 2773.95136 | 3374.84124 | 3617.86452 | 3241.52709 | 3501.67849  |
+
+| | |
+| --- | --- |
+| ![](images/port_max_qps.png) | ![](images/port_max_qps_p.png) |
+
+
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 do you have paragraphs to explain the result listed in the table and figure?

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 7]
```diff
@@ -0,0 +1,718 @@
...
+| ----- | ------ | ------- | ------- | ------- | -- | ----- | ---------- | ---------- | ---------- | ---------- |
+| run 1 | 577.97 | 1205.43 | 1442.56 | 1450.78 | \- | run 1 | 551.633277 | 1144.8677  | 1460.5069  | 1476.50134 |
+| run 2 | 628.21 | 1282.10 | 1435.15 | 1577.70 | \- | run 2 | 695.677398 | 1245.01266 | 1467.48245 | 1724.25347 |
+| run 3 | 658.90 | 1301.53 | 1530.06 | 1622.19 | \- | run 3 | 587.088398 | 1293.09311 | 1537.51299 | 1673.32094 |
+| run 4 | 598.74 | 1155.63 | 1300.60 | 1308.11 | \- | run 4 | 629.011213 | 1028.6955  | 1524.64036 | 1341.9163  |
+| run 5 | 516.99 | 1175.89 | 1262.10 | 1321.07 | \- | run 5 | 560.736288 | 1121.52361 | 1332.64095 | 1465.50503 |
+| run 6 | 539.59 | 1149.98 | 1209.30 | 1331.10 | \- | run 6 | 618.331119 | 1137.34832 | 1312.16707 | 1410.69305 |
+| Avg   | 586.73 | 1211.76 | 1363.29 | 1435.16 | \- | Avg   | 607.08     | 1161.76    | 1439.16    | 1515.37    |
+| Max   | 658.90 | 1301.53 | 1530.06 | 1622.19 | \- | Max   | 695.68     | 1293.09    | 1537.51    | 1724.25    |
+
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md:396]: @yanmo96 Please give one or two paragraphs to explain the data listed in the table and figure.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 8]
```diff
@@ -0,0 +1,718 @@
...
+
+<br />
+
+## API Test, Subnet:  
+
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Rename this title to "Alcor API Test for Subnet creation and list".

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 9]
```diff
@@ -0,0 +1,718 @@
...
+|       | 5 pods     | 10 pods    | 15 pods    | 20 pods    | 25 pods    |
+| ----- | ---------- | ---------- | ---------- | ---------- | ---------- |
+| run 1 | 122.171943 | 192.234644 | 165.971632 | 189.46115  | 170.649043 |
+| run 2 | 213.156499 | 207.863239 | 208.112991 | 237.870493 | 204.961099 |
+| run 3 | 158.115328 | 123.516043 | 131.565865 | 181.897132 | 140.867057 |
+| run 4 | 109.365801 | 108.49676  | 95.6998481 | 117.866189 | 96.2321135 |
+| run 5 | 94.2282891 | 91.5762681 | 85.3390427 | 103.22579  | 83.0531852 |
+| run 6 | 66.536477  | 71.8048504 | 71.0391698 | 89.0149473 | 70.3936466 |
+| Avg   | 127.262389 | 132.581967 | 126.288091 | 153.222617 | 127.692691 |
+| Max   | 213.156499 | 207.863239 | 208.112991 | 237.870493 | 204.961099 |
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Please give one or two paragraphs to explain the results listed in the table and figure.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 10]
```diff
@@ -0,0 +1,718 @@
...
+|       | 5 pods     | 10 pods    | 15 pods | 20 pods    | 25 pods    |
+| ----- | ---------- | ---------- | ------- | ---------- | ---------- |
+| run 1 | 168.864333 | 196.305945 |         | 202.222719 | 194.40302  |
+| run 2 | 119.860926 | 148.726237 |         | 115.141975 | 175.556609 |
+| run 3 | 73.4193208 | 91.39514   |         | 67.2313068 | 102.948376 |
+| run 4 | 49.9883674 | 66.2545772 |         | 47.2278267 | 76.6703555 |
+| run 5 | 37.4873497 | 49.0405038 |         | 41.3234798 | 64.2762938 |
+| run 6 | 29.7931549 | 40.0456381 |         | 35.6744003 | 51.438147  |
+| Avg   | 79.9022419 | 98.6280068 |         | 84.8036178 | 110.882134 |
+| Max   | 168.864333 | 196.305945 |         | 202.222719 | 194.40302  |
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Please give one or two paragraphs to explain the results listed in the table and figure.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 11]
```diff
@@ -0,0 +1,718 @@
...
+|       | 5 pods | 10 pods    | 15 pods    | 20 pods    | 25 pods    |
+| ----- | ------ | ---------- | ---------- | ---------- | ---------- |
+| run 1 |        | 230.72542  | 239.914779 | 212.821755 | 204.047763 |
+| run 2 |        | 113.390071 | 140.163994 | 143.907342 | 132.528615 |
+| run 3 |        | 64.9454037 | 84.8448904 | 90.7973293 | 84.7995649 |
+| run 4 |        | 48.5527157 | 63.9578724 | 67.3949125 | 61.9916036 |
+| run 5 |        | 45.0106555 | 49.4495767 | 50.9884834 | 49.9585688 |
+| run 6 |        | 49.1970909 | 39.2756892 | 39.0478244 | 43.0528196 |
+| Avg   |        | 91.9702261 | 102.934467 | 100.826274 | 96.0631558 |
+| Max   |        | 230.72542  | 239.914779 | 212.821755 | 204.047763 |
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Please give one or two paragraphs to explain the results listed in the table and figure.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 12]
```diff
@@ -0,0 +1,718 @@
...
+|       | 5 pods     | 10 pods    | 15 pods    | 20 pods    | 25 pods    |
+| ----- | ---------- | ---------- | ---------- | ---------- | ---------- |
+| run 1 | 180.997654 | 190.303224 | 211.520134 | 223.21284  | 211.6647   |
+| run 2 | 87.032023  | 98.4634246 | 76.639953  | 103.562334 | 109.248082 |
+| run 3 | 52.1193785 | 61.2469792 | 50.0596896 | 74.3989519 | 71.4504471 |
+| run 4 | 41.2890729 | 62.7838403 | 36.3336839 | 67.9391031 | 52.7183257 |
+| run 5 | 30.0860401 | 34.7104692 | 29.5738334 | 54.4772135 | 49.2410235 |
+| run 6 | 32.0612311 | 32.9007633 | 28.246774  | 40.2831089 | 44.1368398 |
+| Avg   | 70.5975665 | 80.0681168 | 72.0623447 | 93.9789253 | 89.7432363 |
+| Max   | 180.997654 | 190.303224 | 211.520134 | 223.21284  | 211.6647   |
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Please give one or two paragraphs to explain the results listed in the table and figure.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 13]
```diff
@@ -0,0 +1,718 @@
...
+
+<br />
+
+## API Test, VPC:  
+
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Rename the title to "Alcor API Test for VPC creation".

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 14]
```diff
@@ -0,0 +1,718 @@
...
+| run 2 | 97.9381727 |         | 151.075641 | 141.532611 | 135.6119688 |
+| run 3 | 127.521529 |         | 119.350571 | 150.475931 | 138.4461197 |
+| run 4 | 107.651403 |         | 162.683393 | 142.332809 | 125.1218739 |
+| run 5 | 97.4664581 |         | 157.249331 | 106.5709   | 132.83853   |
+| Avg   | 107.644391 |         | 147.589734 | 135.228063 | 133.0046231 |
+| Max   | 127.521529 |         | 162.683393 | 150.475931 | 138.4461197 |
+
+| | |
+| --- | --- |
+| ![](images/vpc_qps.png) | ![](images/vpc_qps_p.png) |
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Please give one or two paragraphs to explain the results listed in the table and figure.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 15]
```diff
@@ -0,0 +1,718 @@
...
+
+It is hard to get a test to pass too, since Nova booting VM will fail from time to time. We have observe VM creation failure from both Nova controller side and Nova compute side on compute hosts.  
+- From compute side, we mainly see the issue of "tap-device not found" error from ACA. Which can be *'fixed'* if we change the Alcor timeout from 300 seconds to 600 seconds. Since Nova will re-try after 300 seconds, and put in the tap-device.  
+- From Nova server side on the controller we have observe many different type of errors when try to create VMs really fast. For example, nova-schedular will have many issues finding the right host. Then messaging services sometimes will error out too. Lastly, the SQL DB on OpenStack cannot handle high concurrent, when we create/delete VMs at a fast speed, the SQL DB will start to have miss match between tables.  
+
+Bellow are a list of different tests we did with VM creation, the results does not seem really ideal. But when compare with others performance on the internet, they do seems to be reasonable with our current OpenStack configuration.  
+
+## End-to-end Test, Small VPC:
+
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Rename the title to "End-to-end VM creation Test for Small VPC scenario" and please give a paragraph to explain what is the small VPC.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 16]
```diff
@@ -0,0 +1,718 @@
...
+| ----- | ---------- |
+| run 1 | 3.34631731 |
+| run 2 | 4.60502671 |
+| run 3 | 4.66211621 |
+| run 4 | 4.67599231 |
+| run 5 | 0.97742867 |
+
+<br />
+
+## End-to-end Test, Big VPC:
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Rename the title to "End-to-end VM creation Test for Large VPC scenario" and please give a paragraph to explain what is the large VPC.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 17]
```diff
@@ -0,0 +1,718 @@
...
+
+![](images/ts3_30c.png)
+
+|       | 5 pods     |
+| ----- | ---------- |
+| run 1 | 3.34631731 |
+| run 2 | 4.60502671 |
+| run 3 | 4.66211621 |
+| run 4 | 4.67599231 |
+| run 5 | 0.97742867 |
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md:653]: @yanmo96 Please give one or two paragraphs to explain the results listed in the table and figure.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 18]
```diff
@@ -0,0 +1,718 @@
...
+}
+```
+
+![](images/ts4_15c.png)
+
+|       | 10 pods    |
+| ----- | ---------- |
+| run 1 | 0.43619856 |
+| run 2 | 0.42073825 |
+| run 3 | 0.43146736 |
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md:718]: @yanmo96 Please give one or two paragraphs to explain the results listed in the table and figure.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 19]
```diff
@@ -0,0 +1,718 @@
...
+
+<br />
+
+## End-to-end Test, Multiple Port per VM:  
+
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md]: @yanmo96 Please rename the title to "End-to-end VM creation Test for Multiple Port per VM scenario" and give a paragraph to explain the scenario.

File: benchmark/AlcorPerformanceTestReport/README.md [hunk 20]
```diff
@@ -0,0 +1,718 @@
...
+    "hooks": [],
+    "sla": {
+        "failure_rate": {
+            "max": 0
+        }
+    }
+}
+```
+
+![](images/alcor-port_9n-10s-10p-10t-10c_run1-2022-03-05_00-41-28.json_boxplot.png)
```
Comment (cj-chung) [inline benchmark/AlcorPerformanceTestReport/README.md:777]: Please give one or two paragraphs to explain the results in the figure.

**Review by cj-chung — CHANGES_REQUESTED** [review]

**Comment by cj-chung** [comment]
@yanmo96 Please go through the report again to fix some grammar and grammatical tense issues. You use Future tense a lot in the report, but actually you've done most of the future tense sentences in the report.

**Comment by yanmo96** [comment]
> @yanmo96 Please go through the report again to fix some grammar and grammatical tense issues. You use Future tense a lot in the report, but actually you've done most of the future tense sentences in the report.

Will make the above changes! Thanks for the feedback!

**Comment by cj-chung** [comment]
@yanmo96 Please put the date or duration you conduct these test in the overview section.

**Comment by yanmo96** [comment]
> 

Hi James, I have added in the changes. Thanks

**Review by cj-chung — APPROVED** [review]
LGTM

**Review by xieus — APPROVED** [review]

**Pull request merged**